### ISSUE TITLE
mapcat, flatmap, bind...

### ISSUE THREAD

**Issue by theleoborges** [weavejester/reagi#5]
I've been playing with reagi and was surprised by the current of behaviour of `mapcat`.

In Clojure it essentially flattens sequences. In reagi however it operates on the `Events` data type so the behaviour I was expecting is one of "flattening" event streams instead of simply flattening lists inside a stream.

To illustrate a little, if the current mapcat function had a type signature, it would be this:

``` haskell
mapcat :: (a -> [b]) -> Events a -> Events b
```

However what I would like is for it to work with the following signature:

``` haskell
mapcat :: (a -> Events b) -> Events a -> Events b
```

Apologies for the type signature in haskell. I'm happy to expand on it if needed. But basically a function working to match such type signature allows us to sequence streams together.

If this issue gets anywhere, one question would be whether mapcat should be changed or a new function, flatmap, should be added. Personally I prefer the former.

Given I found myself wanting `flatmap` when working with Reagi, I started work on a new PR for it but before going too deep down that route I wanted to get feedback on how to best fit this in the design goals of the library.

What follows is a simple implementation of `flatmap` with an example interaction:

``` clojure
(defn- flatmap-ch [f in]
  (let [out (chan)]
    (go-loop []
      (if-let [msg (<! in)]
        (let [stream (f (r/unbox msg))
              stream-ch (chan)]
          (r/sink! stream stream-ch)
          (go-loop []
            (when-let [x (<! stream-ch)]
              (>! out x)
              (recur)))
          (recur))))
    out))

(defn flatmap
  ([f stream]
     (let [ch (#'r/tap stream)]
       (r/events (flatmap-ch f ch) {:dispose #(close! ch), :deps stream}))))

(def t (r/behavior (System/currentTimeMillis)))
(def interval-stream (r/sample 500 t))

;; Usage
(def flatmapped (flatmap (fn [time]
                           (let [c (chan)
                                 e (r/events c)]
                             (go (doseq [n (range 5)]
                                   (>! c n)))
                             e))
                         interval-stream))

(def c (chan))
(r/sink! flatmapped c)
(go-loop []
  (when-let [x (<! c)]
    (prn "Value from channel:" x)
    (recur)))

;; it will output this until c is closed....
;;"Value from channel:" 0
;;"Value from channel:" 1
;;"Value from channel:" 2
;;"Value from channel:" 3
;;"Value from channel:" 4
;; ...

(close! c)
```

This isn't a particularly useful example but it's simple enough to demonstrate the idea. A useful scenario for this is a stream that gets some data from a service and flatmaps itself onto another stream that gets details from another service for each value emitted.

Reagi is young and there isn't much code written in it out there so I'm still trying to find my feet and  working out what usage patterns might look like based on my previous experience using other FRP libraries. I would love some feedback on it.

Finally, if this is really of interest I'd be more than happy to turn this into a proper PR.

**Comment by weavejester** [comment]
Unlike seqs in Clojure, event streams in Reagi have no ending point and therefore cannot be concatenated together. This is why `mapcat` expects a function that returns a collection, rather than an event stream.

The behaviour you describe is sometimes referred to as a `switch`. Instead of concatenating the event streams together, you switch to the latest one. I've been meaning to write it, but I haven't actually run into a situation where I need it. I'd be very interested to know what problem prompted you to want this function.

**Comment by theleoborges** [comment]
I realised that fact - that event streams in reagi had no ending point - and in an earlier implementation of flatmap I used a sentinel value to signal that this stream would not push any more values through. I managed to work around that though as you can see above.

However `switch` isn't really the same thing - as I understand it anyway.

A common example of `switch` is performing a search from a text input field. You're always in interested in the most recent nested stream.

`flatmap` on the other hand keeps itself "subscribed" to the original stream so it keeps sending it's values through, returning new streams and flattening them out into a single stream at the end.

For example:

``` clojure
(def source-stream (r/events))
(def flatmapped (flatmap (fn [n]
                           (let [c (chan)
                                 e (r/events c)]
                             (go (doseq [n1 (range n)]
                                   (>! c n1)))
                             e))
                         source-stream))

(def c (chan))
(r/sink! flatmapped c)
(go-loop []
  (when-let [x (<! c)]
    (prn "Value from channel:" x)
    (recur)))


(r/push! source-stream 3)
(r/push! source-stream 2)

;; This would output
;;"Value from channel:" 0
;;"Value from channel:" 1
;;"Value from channel:" 2
;;"Value from channel:" 0
;;"Value from channel:" 1
```

So we actually get a flattened view of everything, not just the latest.

A real case where I would want to use that is coordinating 3rd party API calls.

In pseudocode:

``` clojure
(-> (resources-stream "accountID")
      (filter "resourceType")
      (flatmap resource-details-stream)
      (map "resouceStatus")
      (subscribe (fn [status]   ;; update screen here....or something else
  )))
```

I hope this example shows in a little more detail the use case I have in mind.

**Comment by weavejester** [comment]
So this is more like a merge, except instead of a fixed set of streams, it's merging across a stream of streams?

Are the semantics effectively identical to:

``` clojure
(r/reduce r/merge stream-of-streams)
```

If so, I'm not sure I completely follow the use-case. It seems that you'd just end up with a stream that gets more and more data pushed through it.

**Comment by theleoborges** [comment]
Strictly speaking it would be closer to:

``` clojure
(defn flatmap [f stream]
  (r/reduce r/merge (r/map f stream)))
```

I say this because the stream of streams if generated based on the function, and not in parallel.

so mapping `f` over `stream` would give us a stream of streams. However what ends up happening is that we only get the latest value of the stream in the go-loop. Also we don't get rid of the last level of nesting so we still get values wrapped in Events.

I like how Rx visualises these operations. Flatmap looks like this in a marble diagram:

![flatmap](https://raw.githubusercontent.com/wiki/Netflix/RxJava/images/rx-operators/flatMap.png)

I find these extremely helpful when taking about stream related things.

**Comment by weavejester** [comment]
Okay, I think we're on the same page, but I still don't quite understand the use-case. You'd just end up with an event stream that keeps using up more and more resources indefinitely.

It sounds as if the use-case you're considering is to use `flatmap` to generate event streams from an I/O source. I'd suggest dropping back to core.async for that, as core.async is already quite good at pulling data from I/O sources, whereas Reagi is largely concerned with functional transformations of data over time.

**Comment by theleoborges** [comment]
You're right. This particular use case does involve IO. 

I didn't get that far in the comments but the idea is to have the IO work done using core.async 's thread macro, pushing into a channel from which I'd create a stream. From there on, it's simply stream processing of values over time. 

I didn't go into much detail for the use case as this is client work. Happy to have a chat in procure though if you'd like to.

**Comment by weavejester** [comment]
If you like, we can discuss your specific use-case over email in confidence. I'm very interested in how  Reagi is best used to solve problems.

**Comment by theleoborges** [comment]
Great. I'm in transit atm with limited connectivity but as soon as I can
I'll send you an email with more details.

**Comment by weavejester** [comment]
After discussions over email and some consideration, I've added a `completed` function that should facilitate functionality like this.

I've been considering a `flatten` and `switch` functions for working with streams of streams, but I haven't settled on names yet.

**Comment by weavejester** [comment]
`flatten` has been added in version 0.9.1.

**Comment by theleoborges** [comment]
Hi @weavejester ,

Apologies for taking me so long to comment on this.

Thanks for adding the `completed` function. This would definitely facilitate implementing `flatmap`. Would you still consider adding it to Reagi. I'm happy to submit a pull request for it.

**Comment by weavejester** [comment]
I think `flatten` plus `map` is good enough for now.

**Comment by theleoborges** [comment]
Fair enough.. That certainly works. 

I'll use my own utility wrapper for now. It simply delegates to `map` and `flatten`.

Tks
